Modify unique index on space names.

VineetReynolds opened the pull request:
The index now checks for uniqueness across owners and space names,
instead of enforcing uniqueness on space names globally.
Fixes #1000

VineetReynolds commented:
[test]

aslakknutsen commented:
What happens to the unique test on spaces with this change: https://github.com/almighty/almighty-core/blob/master/space/space_test.go#L46

It still runs, but maybe we could make the test a bit more explicit on what its testing and why it should fail?

VineetReynolds commented:
@alexeykazakov I've added another test `TestFailCreateSameSpaceNameSameOwner` to address your review comment.

@aslakknutsen I've split the create space unit tests into three separate functions. I'll expand the one on space name validations when I work on subsequent PRs to address validation changes that have been requested.

hectorj2f commented on controller/space_test.go:
```diff
@@ -482,6 +482,60 @@ func (rest *TestSpaceREST) TestListSpacesNotModifiedUsingIfNoneMatchHeader() {
 	test.ListSpaceNotModified(rest.T(), svc.Context, svc, ctrl, nil, nil, nil, &ifNoneMatch)
 }
 
+func (rest *TestSpaceREST) TestSuccessCreateSameSpaceNameDifferentOwners() {
```
cannot we create an assert to validate that the `OwnerId` are different ?

sbose78 commented:
Will merge this after Friday evening demo.

alexeykazakov commented:
@VineetReynolds can you please fix conflict?

aslakknutsen commented:
hmm.. something broke

VineetReynolds commented:
@aslakknutsen The build fails because JSONErrorResponse  in jsonapi_utility returns nil, as the type assertion here fails: https://github.com/almighty/almighty-core/blob/master/jsonapi/jsonapi_utility.go#L128 I'd like pointers on fixing this, as it would affect other callpoints.

hectorj2f commented:
@VineetReynolds Why don't you propagate the error like this ?
```
jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrBadRequest(err.Error()))
        return ctx.BadRequest(jerrors)
```

aslakknutsen commented:
@VineetReynolds Ah, this is not fixed yet. I gave some pointer on chat but I'll repeat:

Remove the default https://github.com/almighty/almighty-core/blob/master/jsonapi/jsonapi_utility.go#L144 and make that the end statement "return nil". That should cover any glitches of unsupported Response types based on found error.. 

Second, you seem to be able to produce something the internals see as a BadRequest, then maybe it's an indication that the endpoint should actually have a BadRequest option in th edesign.

VineetReynolds commented:
@aslakknutsen That turned out to be a no-go. Tried other combinations as well, but they resulted in unit or integration test failures, possibly breaking at least one other functionality. The best hack was to return a error in the controller, but that violates the controller's contract to write to the response always, instead of throwing an error higher up, resulting in valid runtime behavior but a broken integration test.

Nevertheless I can now accurately describe the edge case we're hitting:

1.`jsonapi.JSONErrorResponse` does not return an error. It fires a `ctx.BadRequest` to write a 400 response, but no error is returned. This is sufficient to enforce the controller's contract to return a 400 in some cases.
2. `application.Transactional` examines whether `todo(tx)` has returned an error. From 1, it follows that no error is returned to `Transactional` and hence it doesn't roll back the transaction. It attempts to commit the transaction, and throws another error.
3. Both of these errors are written as separate JSONAPI error responses to the HTTP response writer resulting in an invalid response. If (1) returned an error, then the transaction in (2) is rolled back and only the first error reported in the controller appears in the response, but that requires changing the JSONErrorResponse behavior to always return an error.

I think we need a way to respond with a HTTP error code in the controller, for functions in a transaction context, without throwing an error to force a transaction rollback, or to prevent the trapped errors in the transaction context from bubbling up above the controller boundary.

VineetReynolds commented:
@hectorj2f I can't use `ErrorToJSONAPIErrors` followed by a `ctx.BadRequest` as I need the transaction to be rolled back. Returning `ctx.BadRequest` does not rollback the txn. Return the raw error or the JSONAPI error is also not possible for me, as it is bubbled up by `Transactional` and the controller is forced to return an error unless it handles it.

VineetReynolds commented:
@aslakknutsen I've opted to make a smaller change, to return the error without writing it to the HTTP response writer as a JSONAPI error within the controller. By letting the error bubble upwards, the transaction is rolled back, and eventually the app responds with the error expected by the client:

    {"errors":[{"code":"bad_parameter","detail":"Bad value for parameter 'Name': 'Test space-Dup-test5' (expected: 'unique')","status":"400","title":"Bad parameter error"}]}

I've had to remove the failing test, as there was no way I could use the `test.CreateSpaceXXX` methods without fixing up `application.Transactional` or how it's used.
